source/solvers/post_processors_smoothing.cc

```diff
@@ -0,0 +1,323 @@
+                        (phi_vf[j] * phi_vf[i]) * fe_values.JxW(q);
+                    }
+                }
+            }
+          cell->get_dof_indices(local_dof_indices);
+          constraints.distribute_local_to_global(local_matrix,
+                                                 local_dof_indices,
+                                                 *system_matrix);
+        }
+    }
+  system_matrix->compress(VectorOperation::add);
+}
+
+template <int dim, typename VectorType>
+const TrilinosWrappers::MPI::Vector &
+PostProcessorSmoothing<dim, VectorType>::solve_L2_projection()
+{
+  // Solve the L2 projection system
+  const double linear_solver_tolerance = 1e-15;
+
+  completely_distributed_solution =
+    TrilinosWrappers::MPI::Vector(this->locally_owned_dofs,
+                                  this->mpi_communicator);
+
+  SolverControl solver_control(
+    this->simulation_parameters.linear_solver.max_iterations,
+    linear_solver_tolerance,
+    true,
+    true);
+
+  TrilinosWrappers::SolverCG solver(solver_control);
+
+  //**********************************************
+  // Trillinos Wrapper ILU Preconditioner
+  //*********************************************
+  const double ilu_fill =
+    this->simulation_parameters.linear_solver.ilu_precond_fill;
+  const double ilu_atol =
+    this->simulation_parameters.linear_solver.ilu_precond_atol;
+  const double ilu_rtol =
+    this->simulation_parameters.linear_solver.ilu_precond_rtol;
+
+  TrilinosWrappers::PreconditionILU::AdditionalData preconditionerOptions(
+    ilu_fill, ilu_atol, ilu_rtol, 0);
+
+  std::shared_ptr<TrilinosWrappers::PreconditionILU> ilu_preconditioner =
+    std::make_shared<TrilinosWrappers::PreconditionILU>();
+
+  ilu_preconditioner->initialize(*system_matrix, preconditionerOptions);
+
+
+  solver.solve(*system_matrix,
+               completely_distributed_solution,
+               system_rhs,
+               *ilu_preconditioner);
+
+  constraints.distribute(completely_distributed_solution);
+
+  return completely_distributed_solution;
+}
+
+template <int dim, typename VectorType>
+const TrilinosWrappers::MPI::Vector &
+PostProcessorSmoothing<dim, VectorType>::calculate_smoothed_field(
+  const VectorType &            solution,
+  const DoFHandler<dim> &       dof_handler_fluid,
+  std::shared_ptr<Mapping<dim>> mapping_fluid)
+{
+  generate_mass_matrix();
+  generate_rhs(solution, dof_handler_fluid, mapping_fluid);
+  return solve_L2_projection();
+}
+
+template <int dim, typename VectorType>
+const DoFHandler<dim> &
+PostProcessorSmoothing<dim, VectorType>::get_dof_handler() const
+{
+  return dof_handler;
+}
+
+template class PostProcessorSmoothing<2, TrilinosWrappers::MPI::Vector>;
+template class PostProcessorSmoothing<3, TrilinosWrappers::MPI::Vector>;
+template class PostProcessorSmoothing<2, TrilinosWrappers::MPI::BlockVector>;
+template class PostProcessorSmoothing<3, TrilinosWrappers::MPI::BlockVector>;
+
+template <int dim, typename VectorType>
```

Comment:
I am curious, how will you manage the fact that vorticity has multiple components? (e.g. ti's a vector)

Reply:
I think that I will either way need to calculate the smoothed field 1 dim at a time anyways, is there a way for me to convert a 'TrilinosWrappers::MPI::Vector' to a 'std::vector<double>' ?

Reply:
Makes sense. Let's do this.